neps/nep-0413.md

```diff
@@ -0,0 +1,76 @@
+
+###  The problem with a restricted access key
+
+Originally, NEAR was designed with the assumption that access keys are only used to sign transactions. Therefore, if those keys are used for any other purpose, it is important to make sure that they can't be abused to sign some undesirable transaction.
+
+When using the keys for authentication, another issue that needs to be handled is a replay attack: a malicious app might request the user to authenticate only to authenticate itself to some other app as the user.
+
+If the wallet allows apps to request the user to sign arbitrary data, both attacks would be possible. Thus, some measures are needed to prevent both attacks:
+
+1. To prevent the malicious signing of transactions, a good solution is to make sure that the data to be signed cannot be parsed as a transaction. This is preferred to having a properly formatted transaction with invalid data (such as an invalid method). The precise format of that data should be made a standard, so that, for example, a future change to transaction format will not make it possible to confuse the data for a transaction.
+2. To prevent one app from relaying a signature to a different app, the data to be signed should include the requesting app's identity (e.g. domain name). Again, a well-defined format will facilitate this.
+
+Using an access key restricted in such a way that it cannot send transactions looks like a good temporary solution. It doesn't solve the second problem, though.
+
+Sending the signature in a query string to an arbitrary URL (even within the correct domain) is not secure. Most apps will have some pages that will leak this data in the Referrer header or by some other means. Using URL fragment instead of a query string will improve this (the fragment isn't included in the Referrer), and somehow permitting only a fixed callback URL will improve this even more.
+
+Also, note that NEAR transaction signatures are not plain Ed25519 signatures but Ed25519 signatures of a SHA-256 hash (see [near/nearcore#2835](https://github.com/near/nearcore/issues/2835)). Any protocol that signs anything with NEAR account keys should use the same signature format.
+
+Hence, the final solution we are settling on is to user a restricted payload as shown below.
+
+## Specification
+
+### `verifyOwner`
+
+Signs the message and verifies the owner. Message is not sent to blockchain.
+
+```jsx
+interface VerifyOwnerParams {
+  // The message requested sign. Defaults to `verify owner` string.
+  message?: string;
+   //Account ID used to sign the message. Defaults to the first account
+  signerId?: string;
+  //Public key used to sign the message. Defaults to the public key of the signed in account.
+  publicKey?: PublicKey;
```

Comment:
Why should these be defaults rather than explicitly specified?

Reply:
Actually, there's a mistake here. I just updated it with the new interface from the referenced implementation in Wallet Selector.